processor/src/stack/overflow.rs

```diff
@@ -1,211 +1,414 @@
-    ///   has the effect of restoring the previous context.
-    pub fn set_last_row_addr(&mut self, last_row_addr: Felt) {
-        if last_row_addr != ZERO {
-            // if we are not setting the last row address to ZERO, we can set it only to the
-            // address of the row actually at the top of the table.
-            let last_row_idx = *self.active_rows.last().expect("overflow table is empty");
-            assert_eq!(self.all_rows[last_row_idx].clk, last_row_addr);
+    // PUBLIC MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Pushes a value into the overflow table in the current context.
+    pub fn push(&mut self, value: Felt) {
+        let clk = self.clk;
+
+        // 1. save history
+        if self.history.is_some() {
+            let table_before_push: Vec<Felt> = self
+                .get_current_overflow_stack()
+                .iter()
+                .map(OverflowStackEntry::value)
+                .collect();
+
+            // Note: we do the `history.is_some()` and `history.unwrap()` instead of matching to
+            // satisfy the borrow checker (due to the call to `get_current_overflow_stack()` which
+            // borrows `self` immutably).
+            self.history
+                .as_mut()
+                .unwrap()
+                .save_stack_to_history_before_clk(clk, table_before_push);
         }
-        self.last_row_addr = last_row_addr;
+
+        // 2. push value
+        self.get_current_overflow_stack_mut().push(OverflowStackEntry::new(value, clk));
     }
 
-    // PUBLIC ACCESSORS
-    // --------------------------------------------------------------------------------------------
+    /// Removes the last value from the overflow table in the current context, if any, and returns
+    /// it.
+    pub fn pop(&mut self) -> Option<Felt> {
+        let clk = self.clk;
 
-    /// Returns address of the row at the top of the overflow table.
-    ///
-    /// If the overflow table in the current execution context is empty, this will return ZERO,
-    /// even if the overall overflow table contains values.
-    pub fn last_row_addr(&self) -> Felt {
-        self.last_row_addr
+        // 1. save history
+        if self.history.is_some() {
+            let table_before_pop: Vec<Felt> = self
+                .get_current_overflow_stack()
+                .iter()
+                .map(OverflowStackEntry::value)
+                .collect();
+
+            // Note: we do the `history.is_some()` and `history.unwrap()` instead of matching to
+            // satisfy the borrow checker (due to the call to `get_current_overflow_stack()` which
+            // borrows `self` immutably).
+            self.history
+                .as_mut()
+                .unwrap()
+                .save_stack_to_history_before_clk(clk, table_before_pop);
+        }
```

Comment:
This code seems to be repeated 4 times - why not make it into a separate helper method. For example:
```rust
fn save_stack_to_history(&mut self) {
    let clk = self.clk;
    if self.history.is_some() {
        let table_before_op: Vec<Felt> = self
            .get_current_overflow_stack()
            .iter()
            .map(OverflowStackEntry::value)
            .collect();
        
        self.history
            .as_mut()
            .unwrap()
            .save_stack_to_history_before_clk(clk, table_before_op);
    }
}
```

Reply:
Done